pkg/kubectl/cmd/drain.go

```diff
@@ -385,35 +418,81 @@ func (o *DrainOptions) getPodsForDeletion() (pods []api.Pod, err error) {
 	return pods, nil
 }
 
-// deletePods deletes the pods on the api server
-func (o *DrainOptions) deletePods(pods []api.Pod) error {
-	deleteOptions := api.DeleteOptions{}
+func (o *DrainOptions) deletePod(pod api.Pod) error {
+	deleteOptions := &api.DeleteOptions{}
 	if o.GracePeriodSeconds >= 0 {
 		gracePeriodSeconds := int64(o.GracePeriodSeconds)
 		deleteOptions.GracePeriodSeconds = &gracePeriodSeconds
 	}
+	return o.client.Core().Pods(pod.Namespace).Delete(pod.Name, deleteOptions)
+}
 
-	for _, pod := range pods {
-		err := o.client.Core().Pods(pod.Namespace).Delete(pod.Name, &deleteOptions)
-		if err != nil {
-			return err
-		}
+func (o *DrainOptions) evictPod(pod api.Pod) error {
+	deleteOptions := &api.DeleteOptions{}
+	if o.GracePeriodSeconds >= 0 {
+		gracePeriodSeconds := int64(o.GracePeriodSeconds)
+		deleteOptions.GracePeriodSeconds = &gracePeriodSeconds
+	}
+	eviction := &policy.Eviction{
+		TypeMeta: unversioned.TypeMeta{
+			APIVersion: "policy/v1alpha1",
+			Kind:       "Eviction",
+		},
+		ObjectMeta: api.ObjectMeta{
+			Name:      pod.Name,
+			Namespace: pod.Namespace,
+		},
+		DeleteOptions: deleteOptions,
 	}
+	// Remember to change change the URL manipulation func when Evction's version change
+	return o.client.Policy().Evictions(eviction.Namespace).Evict(eviction)
+}
 
-	getPodFn := func(namespace, name string) (*api.Pod, error) {
-		return o.client.Core().Pods(namespace).Get(name)
+// deletePods deletes the pods on the api server
+func (o *DrainOptions) deletePods(pods []api.Pod) error {
+	if len(pods) == 0 {
+		return nil
 	}
-	pendingPods, err := o.waitForDelete(pods, kubectl.Interval, o.Timeout, getPodFn)
+
+	supportEviction, err := doesSupportEviction(o.client)
 	if err != nil {
-		fmt.Fprintf(o.out, "There are pending pods when an error occured:\n")
-		for _, pendindPod := range pendingPods {
-			cmdutil.PrintSuccess(o.mapper, true, o.out, "pod", pendindPod.Name, false, "")
+		return err
+	}
+
+	if supportEviction {
+		ifTooManyRequests := false
+		var e error
+		for _, pod := range pods {
+			err := o.evictPod(pod)
+			if apierrors.IsTooManyRequests(err) {
+				e = err
+				ifTooManyRequests = true
+				continue
```

Comment:
> How often does 429 happen?

It depends. We cannot make assumption of that.
If some app set PDB's `minAvailable` to be 100%, it will always return 429 when we trying to evict a pod.